### PR TITLE
Make end-to-end learning work

### PR DESCRIPTION
**Pull request by Erasmo1015**

### CODE REVIEW

File: iq_learn/dataset/expert_dataset.py [hunk 1]
```diff
@@ -88,7 +88,10 @@ def __init__(self,
         # apply permutation to cond
         # print("perm:",perm)
         if cond_type!="none":
-            self.conds = [conds["emb"][i] for i in perm]
+            if type(conds["emb"][0])==torch.Tensor:
+                self.conds = [conds["emb"][i].detach().cpu().numpy() for i in perm]
```
Comment (cameron-chen) [inline iq_learn/dataset/expert_dataset.py:92]: does it affect the gradient back propagation?

File: iq_learn/train_iq.py [hunk 2]
```diff
@@ -179,22 +180,36 @@ def main(cfg: DictConfig):
...
+            expert_obs, _, expert_action, __, ___, expert_cond_detached, true_traj_idx = expert_batch
+            unique_idx = []
+            expert_obs_unique = []
+            expert_action_unique = []
+            for i, traj_idx in enumerate(true_traj_idx):
+                if traj_idx not in unique_idx:
+                    unique_idx.append(traj_idx)
+                    expert_obs_unique.append(expert_obs[i])
+                    expert_action_unique.append(expert_action[i])
+            expert_obs_unique = torch.stack(expert_obs_unique, dim=0)
```
Comment (cameron-chen) [inline iq_learn/train_iq.py]: we want conditions to be unique, not obs & act pairs. 

concrete example: 
- expert batch of size 32
- you can have 32 obs-act pairs
- you can have 4 unique conditions corresponding to these pairs

do you think does this make sense

File: iq_learn/encoder/hssm_rl_hil.py [hunk 3]
```diff
@@ -680,16 +683,38 @@ def count_z(one_traj_z):
...
+        self.m_feat = LinearLayer(input_size=2, output_size=self.dist_size*2)
+        # self.transformer = nn.Transformer(d_model=self.dist_size*4, nhead=8, num_encoder_layers=6, num_decoder_layers=6, dim_feedforward=2048, dropout=0.1)
+        self.transformer = ShallowTransformer(
+            self.dist_size*4,
+            seq_len,
+            2,
+            key_hidden_size=self.dist_size*4, 
+            value_hidden_size=self.dist_size*4
+        )
+        self.compact_last = LinearLayer(input_size=self.dist_size*4, output_size=self.dist_size*2)
```
Comment (cameron-chen) [inline iq_learn/encoder/hssm_rl_hil.py]: key_hidden_size (value_hidden_size) coupled with dist_size essentially makes our code loss the flexibilty. 

Can we do key_hidden_size, value_hidden_size separate arugments? say 256.

Something like: 
```
    def instantiate_prob_encoder(self, dist_size=None, seq_len=1000, key_hidden_size=256,value_hidden_size=256):
        '''
        Instantiate the probability encoder
        :param dist_size: size of the distribution, equal to the condition size
        '''
        self.dist_size = dist_size if dist_size is not None else 10
        self.z_logit_feat = LinearLayer(input_size=self.latent_n, output_size=int(key_hidden_size/2))
        self.m_feat = LinearLayer(input_size=2, output_size=int(key_hidden_size/2))
        self.transformer = ShallowTransformer(
            key_hidden_size,
            seq_len,
            2,
            key_hidden_size=key_hidden_size, 
            value_hidden_size=value_hidden_size
        )
        self.compact_last = LinearLayer(input_size=key_hidden_size, output_size=dist_size*2)

        self.prob_encoder = True
        return self.prob_encoder
```

**Comment by cameron-chen** [comment]
can you show me in which line of code you generate a condition sample from the encoder and give it to the decoder?

this aims to confirm one thing: whether or not the gradient can back prop from BC loss (**reconstruction loss**) to the encoder.

File: iq_learn/dataset/expert_dataset.py (hunk 1 above)
Comment (Erasmo1015) [reply iq_learn/dataset/expert_dataset.py:92]: No. The probabilistic encoder generate cond and dist_params only using the traj index from buffer instead of directly the cond from buffer.

File: iq_learn/train_iq.py (hunk 2 above)
Comment (Erasmo1015) [reply iq_learn/train_iq.py]: Changed to the following:
- expert batch of size 32
- 32 obs-act pairs
- 32 conditions corresponding to these pairs

File: iq_learn/encoder/hssm_rl_hil.py (hunk 3 above)
Comment (Erasmo1015) [reply iq_learn/encoder/hssm_rl_hil.py]: Suggestion implemented

File: iq_learn/train_iq.py [hunk 4]
```diff
@@ -181,35 +181,23 @@ def main(cfg: DictConfig):
...
             dist_params_list = emb_list["dist_params"]
-            dist_params_list = [dist_params_list[i] for i in unique_idx]
             mu = [dist_params[0] for dist_params in dist_params_list]     
             log_var = [dist_params[1] for dist_params in dist_params_list]
             mu = torch.stack(mu, dim=0)
             log_var = torch.stack(log_var, dim=0)
            
             expert_cond_list = emb_list["emb"]
-            expert_cond_list = [expert_cond_list[i] for i in unique_idx]
             expert_cond = torch.stack(expert_cond_list, dim=0)
```
Comment (Erasmo1015) [inline iq_learn/train_iq.py:202]: The cond from encoder goes here. 
1. cond is forwarded by encoder at each training step
2. cond is passed to agent.bc_update() for forwarding through the actor
3. agent.bc_update() calculate loss afterwards

**Review by cameron-chen — APPROVED** [review]
